pandas/core/algorithms.py

```diff
@@ -407,6 +407,28 @@ def unique(values):
     return unique_with_mask(values)
 
 
+def nunique_ints(values: ArrayLike) -> int:
+    """
+    Return the number of unique values for integer array-likes.
+
+    Significantly faster than pandas.unique for long enough sequences.
+    No checks are done to ensure input is integral.
+
+    Parameters
+    ----------
+    values : 1d array-like
+
+    Returns
+    -------
+    int : The number of unique values in ``values``
+    """
+    if len(values) == 0:
+        return 0
+    values = _ensure_data(values)
+    result = (np.bincount(values.ravel().astype("intp")) != 0).sum()
```

Comment:
On a 32-bit machine, without the `astype("intp")` numpy will fail (can't safely cast to int32); with it however there could be issues if values exceed the 32bit limitation. This provides a decent speedup, but could be removed if it might cause issues.

Reply:
Curious what call is trying to cast to int32 before the `astype`, the `bincount` or `!=`?

Reply:
bincount - I verified this in https://github.com/pandas-dev/pandas/pull/49652/commits/f3a3ebba9519b5a7eb942c0f060a52a0bed72f5a. The implementation appears to require intp, but I don't know the details as to why. There are a few issues on the NumPy tracker in this regard, but they appear to all be about unsigned integers rather than this case.

Reply:
Gotcha. Would be good to add a comment here noting that bintcount is requiring the casting to intp.